Try adding a 'spotlight mode' type effect when template part or child is selected.

Addison-Stavlo opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
<!-- Please describe what you have changed or added -->
Re-exploring the idea of adding a 'spotlight mode' effect on template parts.
Here, we add a new class to pass down from the block-list if a template part or any of its children are selected.  And some style rules similar to the current focus-mode.

This works as a default behavior for template part/child selection and the styles are overridden if 'spotlight mode' is enabled.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->
![spotlight-mode-template-part-2](https://user-images.githubusercontent.com/28742426/94323814-3a1d4d00-ff65-11ea-8696-cb109634e220.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

Addison-Stavlo committed:
apply mode from blocklist + css

Addison-Stavlo commented:
Any thoughts on how this feels @shaunandrews @jameskoster ?  It will be overridden if 'spotlight mode' becomes enabled, but I don't think thats necessarily a bad thing.

dubielzyk commented:
I quite like it. We're not communicating why we're entering spotlight mode to the user, but perhaps that's fine. Would this also trigger if you have nested template parts?

jameskoster commented:
I think spotlight is the way to go, but the finer details of what happens when inner blocks are selected, or when multiple instances of a single template part exist on the canvas, needs some refinement.

In #25452 I worked on a couple of crude prototypes that we can perhaps discuss here.

In this [latest iteration](https://cdpn.io/jameskoster/debug/XWdGWBr):

* The selected template part is outlined and spotlighted. The outline helps when spotlight mode is enabled.
* Any other instances of that template part are outlined but excluded from the spotlight.
* When an inner block is selected the entire parent template part remains spotlighted and equivalent inner blocks in other instances of the same template part are also spotlighted and outlined.

Addison-Stavlo commented:
> Would this also trigger if you have nested template parts?

@dubielzyk - Good question, it is a little odd right now the nested template parts are spotlighted together regardless of whether the parent or child is selected.  I will try to make this better so that only the child is spotlit when it is the one selected.

> when multiple instances of a single template part exist on the canvas, needs some refinement.

@jameskoster  - I think it is fair to follow up and reiterate regarding how to handle duplicates of the same template part that aren't selected.  Currently, it isn't possible to have 2 of the same template part on one page and that bug is in the process of being worked out technically.  But should that prevent us from adding the ability to spotlight for the selected template part in the meantime?  If we agree that the opacity changes are the direction forward, then a small scoped change like this would be a good first step.

> * The selected template part is outlined and spotlighted. The outline helps when spotlight mode is enabled.
> * Any other instances of that template part are outlined but excluded from the spotlight.
> * When an inner block is selected the entire parent template part remains spotlighted and equivalent inner blocks in other instances of the same template part are also spotlighted and outlined.

The first outline already exists as a subtle gray border and I agree it does work well with the template part spotlight mode.  The second would be a simple change in the blocks editor files, and the third would require some changes in the block-list/wrappers but should be doable as well if we determine that is the way to go.  Both of these I could see as potential follow ups after we are able to successfully render 2 of the same template parts on one page.

jameskoster commented:
> Both of these I could see as potential follow ups after we are able to successfully render 2 of the same template parts on one page.

Seems fair to me :)

shaunandrews commented:
This feels nice. If we can make the logic work for nested template parts, I'd love to see this merged.

Addison-Stavlo committed:
added class to most active template part

Addison-Stavlo committed:
css rules to support nested/active template parts

Addison-Stavlo commented:
> This feels nice. If we can make the logic work for nested template parts, I'd love to see this merged.

Nice!  848b797 + dcc2d5f should have this working for the nested cases now. 😁

noahtallen commented:
Nice, the nested case works well. I also appreciate how "spotlight mode" takes precedent over this, so it still works the same as before with template part blocks.

noahtallen commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -114,7 +138,9 @@ function BlockList(
 								'is-dropping-horizontally':
 									isDropTarget &&
 									orientation === 'horizontal',
+								'template-part-highlighting': activeTemplatePartId,
```
Will this work if `activeTemplatePartId` has ID 0?

Addison-Stavlo replied:
The ID generally a string, so `"0"` would be truthy.  If it were a numeric value, it would not work.  However this seems a bit different from the standard convention for blockIds, but we could account for that if it seems necessary?

noahtallen replied:
Gotcha! I didn't realize it was a string here :)

noahtallen commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -46,16 +48,37 @@ function BlockList(
+		const multiSelectedBlockClientIds = getMultiSelectedBlockClientIds();
+
+		// Determine if there is a template part block to highlight.
+		const activeTemplatePartId = ( function () {
```
It feels like this could be expensive to run all the time when template parts don't exist. I wonder if it would be possible to avoid running this code when the template part block has not been registered (That would keep it out of non-fse sites for now)

Addison-Stavlo replied:
> It feels like this could be expensive to run all the time when template parts don't exist.

I didn't think this would be very expensive?  And its up at the block-list level so its not running for every block.

> I wonder if it would be possible to avoid running this code when the template part block has not been registered

Yeah, I think we could gate it behind the experiment being enabled.

noahtallen replied:
> I didn't think this would be very expensive

I guess it'd just be expensive for deeply nested blocks, which would have to parse all the way up the block tree to get a result. But I think it's memoized so that helps. I also didn't realize it wasn't running for every block which is nice :) So it's probably me just being overly cautious. Definitely fine to try out in the site editor 👍

Addison-Stavlo replied:
Gated this to check for the site editor experiment - bea6a37

noahtallen commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -85,6 +85,21 @@
+		transition: opacity 0.1s linear;
+		@include reduce-motion("transition");
+
+		&.is-selected,
```
Can we remove some of these classnames since we can assume that `is-active-template-part` means that it's selected?

Addison-Stavlo replied:
Ah yeah we probably don't need that one anymore 😁

Addison-Stavlo replied:
Looking again, I don't think we can remove any of these selectors and still have it functioning properly.

We could replace the `&.is-selected` with `&.is-active-template-part` if that seems more ideal to you?

noahtallen replied:
I personally think that makes sense, since only the active template part would receive opacity: 1, but it's up to you. Since it's functionally important we don't have to change it

Addison-Stavlo replied:
updated

noahtallen reviewed (commented):
Nice! I like how this works.

before merging, i'd love to see if we could disable it for non-FSE sites (so we can test it out without affecting plugin consumers).

Besides that, from a code point of view, I dislike how we are tightly coupling one block type with the general block list. I wonder if there's a better way to handle it -- perhaps adding `is-active-template-part` to the template part block's edit method (or just adding the right CSS properties to the template part's edit div)

I can see how `template-part-highlighting` needs to be more general though. I'm not thinking of any alternatives to that approach off the top of my head though.

Addison-Stavlo commented:
> I dislike how we are tightly coupling one block type with the general block list. I wonder if there's a better way to handle it 

I understand that concern.  However, we do need to access this area to handle adding the `template-part-highlighting` class to blocks in the editor.  So using this as the place to add `is-active-template-part` consolidates the code for this functionality to one area.  Similar to how 'spotlight mode' adds `is-focus-mode` and `is-focused` here, we are adding the equivalent versions of that for template parts.

While it is 'just one block type', it could be extended to more block type if desired in the future (post content, etc.).  So maybe rebranding these to more generic terms than 'template parts' would make sense? 🤷‍♀️ 

> -- perhaps adding `is-active-template-part` to the template part block's edit method 

We could do that, although it would separate the code used to achieve this functionality into 2 separate packages which to me seems much less ideal.

> (or just adding the right CSS properties to the template part's edit div)

Im not sure I follow what these right CSS properties would be?  Opacity is a bit of a complex case in that it bleeds down, so just setting `opacity: 1` on the template part won't do much if it has a parent already set down to 0.5.  The styles to achieve this opacity reduction without multiplying the effect on nested blocks is very well bundled together at the moment.  If we move all the styles to template part edit, it doesn't make much sense for the `.template-part-highlighting` to be defined in one package and have its styles defined in another.  If we decouple the styles and move them across 2 different packages, that makes this much more difficult to manage in the future.

In the end, using the block-list / wrappers seems necessary for implementing an editor wide content opacity control and while we may be referencing 'one specific block type' in the general list component that seems much more manageable than splitting up the functionality and style rules between multiple packages.

noahtallen commented:
Thanks for the detailed explanation!

> So maybe rebranding these to more generic terms than 'template parts' would make sense

I would be in favor of this then!

Addison-Stavlo committed:
gate template part check behind site editor experiment

Addison-Stavlo commented:
> > So maybe rebranding these to more generic terms than 'template parts' would make sense

> I would be in favor of this then

Any ideas on names? 😆   `is-active-entity`/`entity-highlighting` ? or instead of 'entity' other ideas like content, context, region... 🤔

noahtallen commented:
ideas:

- `is-active-entity`/`has-active-entity` (changing the second one) 
- `is-active-block-area`/`has-active-block-area`
- `is-active-controller`/`has-active-controller` (like inner block controller)

Addison-Stavlo committed:
rebrand/refactor to support more entity blocks if necessary

Addison-Stavlo commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -533,7 +535,12 @@ export const getBlockParentsByBlockName = createSelector(
 					id,
 					name: getBlockName( state, id ),
 				} ) ),
-				{ name: blockName }
+				( { name } ) => {
+					if ( Array.isArray( blockName ) ) {
+						return blockName.includes( name );
+					}
+					return name === blockName;
+				}
```
Stealing all this from @noahtallen 's nice addition on #25547 that never went through for other reasons.  Its definitely helpful. 😁

youknowriad replied:
what is this change about?

youknowriad replied:
Let's a unit test to this selector to cover that

Addison-Stavlo replied:
updated

Addison-Stavlo reviewed (commented):
I updated this to be capable of easily adding other entity blocks in the future if necessary (such as 'core/post-content'), and updated the classnames to reflect this appropriately `is-active-entity`/`has-active-entity`.  (Nitpicking `has-active-entity` makes me think that the active entity would be a child of the element that has this class on it, when it is really just somewhere in the editor.  Although, this is better than other suggestions and I can't think of anything better.)

noahtallen commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -22,6 +22,8 @@ import useBlockDropZone from '../use-block-drop-zone';
  */
 const BLOCK_ANIMATION_THRESHOLD = 200;
+// List of blocks to 'spotlight' when active in Full Site Editing.
+const ENTITY_AREAS = [ 'core/template-part' ];
```
I do like how this feels more extensible without adding much extra programming complexity

youknowriad replied:
I don't like this personally, it's static and references blocks that might or might not exist in the available blocks. Can't we rely on the "controlled" inner blocks instead, we have a flag for that I believe.

Addison-Stavlo replied:
> might or might not exist in the available blocks

I guess that is one reason running the code that uses this is gated behind the FSE experiment being active for now.

> Can't we rely on the "controlled" inner blocks instead, we have a flag for that I believe.

If we only want to test this for template parts currently, and have the ability to test/add it with other entity controller blocks one at a time, evaluating based on controlled inner blocks would be a bit too general?

Addison-Stavlo replied:
If we are moving the mapSelect logic to a selector, Im thinking this list of allowed blocks to enable this feature will go there with it.

youknowriad replied:
> evaluating based on controlled inner blocks would be a bit too general?

Right now, the only "controlled" blocks are "Post content" or "Template part" and I believe later we'll have "Reusable blocks". All these blocks seem like good candidates for this behavior for me.

noahtallen replied:
Checking if it's an inner block controller does make sense to me. I personally haven't been sure how far we should rely on that for this information, but I do like the simplicity.

The only challenge would be if a 3rd party user creates a block which is a controller, but doesn't expect this UI side effect. I don't think that needs to be a big concern for now, though.

Addison-Stavlo replied:
Regarding detecting by controlled inner blocks - while I do think that would clean things up and that the other controlled inner blocks are good candidates for this behavior, I don't think the design discussions have gotten much of a chance to think about this.

Moreover, detecting by a list of allowed blocks allows us to easily re-use the new selector in other areas that have different needs.  For example, I cleaned up the `packages/edit-site/src/components/header/document-actions/index.js` `useSecondaryText` hook quite a bit by being able to reuse this logic.  In that component, it was argued to only show the title for active template parts and not post content.

So if we wanted to try this spotlight behavior for post content we could just update the selection to be
`getActiveBlockIdByBlockNames( [ 'core/template-part', 'core/post-content' ] )`, yet still have the previously mentioned area just using `getActiveBlockIdByBlockNames( [ 'core/template-part' ] )` (or vice versa).

This format also allows for future usage outside controlled blocks in general.  The new selector could be used for other blocks such as group block, columns block, or whatever else might be useful in future development.

youknowriad replied:
I still disagree about having this list in `block-editor` personally. If this is to be a curated list, I think it needs to be a setting of `BlockEditorProvidder` but to be honest, I don't see why design would not want post content to not use spotlight mode like template part, in some ways, it's just a "content" template part. 

If you prefer we can start with an unstable block editor setting?

Addison-Stavlo replied:
> If you prefer we can start with an unstable block editor setting?

I think that would definitely be better than what we are currently doing.  Would that be to define the list [in the store defaults](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/defaults.js) with an `__experimental` prefix?

> I don't see why design would not want post content to not use spotlight mode like template part, in some ways, it's just a "content" template part.

I agree, maybe we can update this from a list in settings to checking for controlled inner blocks in the future instead if this is desirable.  

@shaunandrews @kellychoffman @jameskoster @dubielzyk @olaolusoga - Has there been much discussion or thought regarding using this sort of spotlight mode for Post Content (and in the future reusable blocks) as well?  Similarly with what shows up as the secondary/selected item in the document-actions name panel in the top bar?  I get the feeling that there should be consistency between what we show in the top-bar and what we highlight with the 'spotlight' mode.

Addison-Stavlo replied:
@youknowriad - Updated to pass the list into settings from the site editor's initialization, with a default empty array on the block-editors side.  The selector will also return early if there is nothing in that array so we got rid of the check for the experiment being active as well. (thanks to some help from @noahtallen 🎉 )

noahtallen reviewed (approved):
nice! I think this works well, and it the changes make sense to me.

Addison-Stavlo commented:
Great!  We could use a core team member to approve since we are making these changes in the block list/wrapper (although still gated behind the FSE experiment being active) as well as updating that selector.  

@youknowriad - if you get some time spare to sanity check this one.

youknowriad commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -46,16 +51,47 @@ function BlockList(
+
+		// Determine if there is an active template part (or other entity) to spotlight.
+		// Restrict to Full Site Editing experiment.
+		const isFullSiteEditingEnabled = getSettings()
```
Why are we restricting to FSE here? I can see a world where this is useful for reusable blocks if they are refactored to work like template parts.

Addison-Stavlo replied:
Mostly out of caution for the time being as the only block currently being tested in this is the template part.  But I agree, this could definitely be useful for non-FSE things.  Its just that the designs that have driven this to life have only thought about template parts for now.

Addison-Stavlo replied:
Now it is just there because the only block in the list is of allowed blocks is only registered in FSE, so it stops the selector from running unnecessarily.  We could remove this if desired, or if other blocks are added to this functionality.

noahtallen replied:
I think the selector still has to do the same "expensive" calculations to figure out if a block is not activee

Addison-Stavlo replied:
> I think the selector still has to do the same "expensive" calculations to figure out if a block is not active

Ah gotcha.  It should never change either, so maybe we run it once to get the value on mount or something?  Or... just get rid of it?

Addison-Stavlo replied:
Or maybe it would make sense to make an `__experimentalIsFullSiteEditingActive` selector that is memoized? 🤔

Addison-Stavlo replied:
> I think the selector still has to do the same "expensive" calculations to figure out if a block is not active

Im not sure I follow.  `getSettings` ? Or I guess the new selector since you are mentioning the block being active? Luckily we aren't calling it on/for every block, just on the block list and only for the selected block. 🤔

noahtallen replied:
> Now it is just there because the only block in the list is of allowed blocks is only registered in FSE, so it stops the selector from running unnecessarily

I actually misunderstood what you were saying! I read this as "the list of allowed blocks stops the selector from running unnecessarily because the allowed blocks are only registered when FSE is active". So you can disregard my comment 😅

youknowriad commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -46,16 +51,47 @@ function BlockList(
+					return last( entityAreaParents );
+				}
+				return null;
+			} )();
```
This is a bit too complex for `mapSelectToProps`, should this be moved to a selector?

Addison-Stavlo replied:
I was just thinking this as well last night, that makes sense to me.

Addison-Stavlo replied:
Also noting moving this logic to a selector will allow us to clean up the document-actions header component in edit-site since it could use this information as well.

Addison-Stavlo replied:
updated

youknowriad commented:
did we try to see how this combines with the global spotlight mode?

Addison-Stavlo commented:
> did we try to see how this combines with the global spotlight mode?

Yup!  global spotlight mode overrides this, which seems to make sense and has good behavior.  So if you are working in an active template part, the entire template part is spotlit.  When you turn on global spotlight mode, only the selected block is spotlit as expected.

Addison-Stavlo committed:
move logic to selector

Addison-Stavlo committed:
change selector name and add doc comments

Addison-Stavlo committed:
update tests for updated selector

Addison-Stavlo committed:
add test for new selector

Addison-Stavlo committed:
make test state reusable for updated parents tests

Addison-Stavlo committed:
cleanup document-actions useSecondaryText to use new selector

Addison-Stavlo committed:
add dependency to selector for selectionEnd

Addison-Stavlo committed:
remove unnecessary style selectors

Addison-Stavlo commented:
Updated this to:
* Create a new selector.
    * use this selector to clean up `useSecondaryText` in document-actions.
*  Add tests for the updated and new selectors.
*  Updated some redundant or unnecessary style rules.

could use another look and testing.

noahtallen commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1798,3 +1805,42 @@ export function isBlockHighlighted( state, clientId ) {
+		}
+		return null;
+	},
+	( state ) => [ state.selectionStart.clientId, state.selectionEnd.clientId ]
```
Not sure if this is needed, but I think we could add `valildBlockNames` to the list here just in case.

Addison-Stavlo replied:
Added. ✅    -  Yeah, probably not needed... but if someone may find an interesting use for this selector that requires that.

noahtallen commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -21,49 +20,18 @@ function getBlockDisplayText( block ) {
+	if ( activeEntityBlockId ) {
 		return {
-			label: getBlockDisplayText( closestParent ),
+			label: getBlockDisplayText( getBlock( activeEntityBlockId ) ),
```
nice, this is awesome :D

noahtallen commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1798,3 +1805,42 @@ export function isBlockHighlighted( state, clientId ) {
+ *
+ * @return {string} The matching block's clientId.
+ */
+export const getActiveBlockIdByBlockNames = createSelector(
```
I'm not a huge fan of this name, but I can't think of anything better either 🙃 Maybe we should mark it as __experimental so that people don't start using it in the meantime?

Addison-Stavlo replied:
Agreed, and sure that makes sense to me!  done

noahtallen commented:
it also is still testing great for me locally 👍

Addison-Stavlo committed:
make selector experimental

Addison-Stavlo committed:
refactor list of blocks to editor settings

Addison-Stavlo committed:
selector return early if no names are passed

noahtallen reviewed (approved):
I like the new changes. Setting that via settings seems to be a lot cleaner, and it helps us avoid running any of the (theoretically expensive?) code outside of the site editor for now.

It also works very well for me locally:

![2020-10-02 12 00 18](https://user-images.githubusercontent.com/6265975/94960205-52cabd00-04a7-11eb-8dbe-db5da36a2519.gif)

The only thing I'm less sure about is the names of the selectors and state we've added, but since we've marked them as experimental, we have time to improve them if folks have better ideas :)

vindl commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -85,6 +85,21 @@
 		}
 	}
+
+	// Active template part highlighting.
```
We could update this to refer to spotlight entities and mention template parts as one specific example.

vindl commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -38,14 +38,21 @@ function BlockList(
 
+		// Determine if there is an active entity area to spotlight.
+		const activeEntityBlockId = __experimentalGetActiveBlockIdByBlockNames(
+			getSettings().__experimentalSpotlightBlocks
```
Would it make sense to make this a default parameter value if it's not passed? Then we could also probably shorten the name of the selector to `__experimentalGetActiveEntityBlockId`, which better aligns it with surrounding variable names that we are using. I guess it depends on how often we anticipate this selector will be used outside of this functionality.

Also I'd probably rename `__experimentalSpotlightBlocks` to `__experimentalSpotlightEntityBlocks` to avoid potential confusion with regular spotlight feature.

Addison-Stavlo replied:
> Would it make sense to make this a default parameter value if it's not passed?
> I guess it depends on how often we anticipate this selector will be used outside of this functionality.

Yeah, I guess that 100% depends on how this selector might be used.  For our purposes, that would make sense!  However, without that change this selector is more general and could be used for other purposes.  Whether it would be, I really don't know. 🤷‍♀️ 

> Also I'd probably rename __experimentalSpotlightBlocks to __experimentalSpotlightEntityBlocks to avoid potential confusion with regular spotlight feature.

👍

vindl reviewed (approved):
I really like the latest round of updates to this, the code looks much more polished now! ✨ Gave it another spin and everything still works as expected on my end. I also made sure that these changes are not affecting the post editor.

I added a few optional naming suggestions.

Addison-Stavlo commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -512,14 +512,16 @@ export const getBlockParents = createSelector(
- * @param {boolean} ascending Order results from bottom to top (true) or top to bottom (false).
+ * @param {Object}          state     Editor state.
+ * @param {string}          clientId  Block from which to find root client ID.
+ * @param {string|string[]} blockName Block name(s) to filter.
```
🤔   Im not sure tbh.  The original use is a string of a single blockName and this is still supported.  Since it can either be a string or an array of strings, whether we use `blockName` or `blockNames` it will still read a bit funky somewhere. ( like `name === blockNames` 🤷‍♀️

Addison-Stavlo committed:
update css comment from template part to entity

Addison-Stavlo committed:
rename the settings list to include 'entity'

Addison-Stavlo committed:
Merge branch 'master' into try/spotlight-mode-for-template-parts

Addison-Stavlo committed:
Merge branch 'master' into try/spotlight-mode-for-template-parts

Addison-Stavlo committed:
Merge branch 'master' into try/spotlight-mode-for-template-parts

Addison-Stavlo committed:
Merge branch 'master' into try/spotlight-mode-for-template-parts

Addison-Stavlo committed:
remove unused import

Addison-Stavlo committed:
Merge branch 'master' into try/spotlight-mode-for-template-parts

vindl reviewed (approved):
LGTM! 🚀
